cpp/dolfinx/fem/interpolate.h

```diff
@@ -107,8 +147,495 @@ void interpolate_from_any(Function<T>& u, const Function<T>& v)
   assert(v.function_space()->mesh());
   if (mesh->id() != v.function_space()->mesh()->id())
   {
-    throw std::runtime_error(
-        "Interpolation on different meshes not supported (yet).");
+    const int nProcs = dolfinx::MPI::size(MPI_COMM_WORLD);
+    const auto mpi_rank = dolfinx::MPI::rank(MPI_COMM_WORLD);
+
+    const int tdim = u.function_space()->mesh()->topology().dim();
+    const auto cell_map
+        = u.function_space()->mesh()->topology().index_map(tdim);
+    const int num_cells = cell_map->size_local() + cell_map->num_ghosts();
+    std::vector<std::int32_t> cells(num_cells, 0);
+    std::iota(cells.begin(), cells.end(), 0);
+
+    // Collect all the points at which values are needed to define the
+    // interpolating function
+    const xt::xtensor<double, 2> x = fem::interpolation_coords(
+        *u.function_space()->element(), *u.function_space()->mesh(), cells);
+
+    // This transposition is a quick and dirty solution and should be avoided
+    auto x_t = xt::zeros_like(xt::transpose(x));
+    for (decltype(x.shape(1)) i = 0; i < x.shape(1); ++i)
+    {
+      for (decltype(x.shape(0)) j = 0; j < x.shape(0); ++j)
+      {
+        x_t(i, j) = x(j, i);
+      }
+    }
+
+    dolfinx::geometry::BoundingBoxTree bb(*v.function_space()->mesh(), tdim,
+                                          0.0001);
+    auto globalBB = bb.create_global_tree(MPI_COMM_WORLD);
+
+    std::vector<std::vector<int>> candidates(x_t.shape(0));
+    for (decltype(x_t.shape(0)) i = 0; i < x_t.shape(0); ++i)
+    {
+      const auto xp = x_t(i, 0);
+      const auto yp = x_t(i, 1);
+      const auto zp = x_t(i, 2);
+      candidates[i]
+          = dolfinx::geometry::compute_collisions(globalBB, {xp, yp, zp});
+    }
+
+    std::vector<std::vector<double>> pointsToSend(nProcs);
```

Comment:
You should use reserve on all the sub-vectors, otherwise the push back in the next loop is going to be quite expensive. I would suggest that each sub-vector reserves the size `3*candidates.size()`

Reply:
Is 3 an heuristic guess or did you mean `nProcs`?

Reply:
3 because you send 3 points, and candidates.size as it is the maximum number of possible points

Reply:
Ah, I understand what you mean now. That would mean allocating enough memory to send each point to every process, right?

Reply:
Yes, but most of it might be unused. We should at least reserve some memory, as using push back does several reserves.

Reply:
It is supposed to be logarithmically amortised, but I don't mind, this is an implementation detail that can always be changed. I wonder if anyone else will chip in into the discussion :thinking: